src/receive_imf.rs

```diff
@@ -1632,86 +1637,136 @@ async fn apply_group_changes(
         return Ok(None);
     }
 
-    let mut recreate_member_list = false;
     let mut send_event_chat_modified = false;
-
+    let mut removed_id = None;
     let mut better_msg = None;
-    let removed_id;
+
+    let self_added =
+        if let Some(added_addr) = mime_parser.get_header(HeaderDef::ChatGroupMemberAdded) {
+            if let Ok(self_addr) = context.get_primary_self_addr().await {
+                self_addr == *added_addr
+            } else {
+                false
+            }
+        } else {
+            false
+        };
+
+    let recreate_member_list =
+        // if we're not in the group, reject don't bother about group members
+        if !chat::is_contact_in_chat(context, chat_id, ContactId::SELF).await? && !self_added {
+            false
+        } else {
+            let apply_group_changes = chat_id
+                .update_timestamp(context, Param::MemberListTimestamp, sent_timestamp)
+                .await?;
+
+            // always reject old group changes
+
+            if apply_group_changes {
+                // Recreate member list if the message comes from a MUA as these messages
+                // do _not_ set add/remove headers
+                if !mime_parser.has_chat_version() && apply_group_changes {
+                    true
+                } else {
+                    match mime_parser.get_header(HeaderDef::InReplyTo) {
+                        // If we don't know the referenced message, we missed some messages which could be add/delete
+                        Some(reply_to) if rfc724_mid_exists(context, reply_to).await?.is_none() => true,
+                        Some(_) => self_added,
+                        None => false,
+                    }
+                }
+            } else {
+                false
+            }
+        };
+
     if let Some(removed_addr) = mime_parser
         .get_header(HeaderDef::ChatGroupMemberRemoved)
         .cloned()
     {
         removed_id = Contact::lookup_id_by_addr(context, &removed_addr, Origin::Unknown).await?;
-        recreate_member_list = true;
-        match removed_id {
-            Some(contact_id) => {
+        if let Some(contact_id) = removed_id {
+            let apply_group_changes = chat_id
+                .update_timestamp(context, Param::MemberListTimestamp, sent_timestamp)
+                .await?;
+            if apply_group_changes {
+                // remove a single member from the chat
+                if !recreate_member_list {
+                    chat::remove_from_chat_contacts_table(context, chat_id, contact_id).await?;
+                    send_event_chat_modified = true;
+                }
+
                 better_msg = if contact_id == from_id {
```

Comment:
If we recreate the member list, better to show appropriate messages for all changed members. Otherwise a shown info is incomplete and a user will see that the contact is removed f.e. but actually remains in the group

Reply:
Would be nice to have that, but atm we can only return _one_ `better_msg` so there is no way to reflect multiple changes without introducing a new stockstring or rewriting the other logic to create multiple system messages. I would rather leave that work to another PR because this one is complex enough and long open aswell.

Reply:
Then if we recreate the member list and can't add all necessary messages in the current code, would be good to just add "Some members were added/removed" message. But maybe do that only if there are multiple additions/removals, looks like it's not a complex check

Reply:
This change needs a new strock string, so I would rather do that in a future pr.

Reply:
This PR here is supposed to fix improve group membership consistency, and it's already pretty big. I think there is no advantage in also improving info messages in the same PR; having separate PRs makes it a lot easier to review & discuss.